src/cargo/ops/resolve.rs

```diff
@@ -325,3 +321,118 @@ fn get_resolved_packages<'a>(resolve: &Resolve,
+    // further refine this `keep` function with *all transitive dependencies* of
+    // the packages we're not keeping. For example consider a case like this:
+    //
+    // * There's a crate `log`
+    // * There's a crate `serde` which depends on `log`
+    //
+    // Let's say we then run `cargo update -p serde`. This may *also* want to
+    // update the `log` dependency as our newer version of `serde` may have a
+    // new minimum version required for `log`. Now this isn't always guaranteed
+    // to work. What'll happen here is we *won't* lock the `log` dependency nor
+    // the `log` crate itself, but we will inform the registry "please prefer
+    // this version of `log`". That way if our newer version of serde works with
+    // the older version of `log`, we conservatively won't update `log`. If,
+    // however, nothing else in the dependency graph depends on `log` and the
+    // newer version of `serde` requires a new version of `log` it'll get pulled
+    // in (as we didn't accidentally lock it to an old version).
+    //
+    // Ok but that's also not the entire story! Updates to the dependency graph
+    // can come from two locations, the `cargo update` command or the manifest
+    // itself. For example the manifest on the filesystem may have been updated
+    // to have an updated version requirement on `serde`. In this case both
+    // `keep(serde)` and `keep(log)` return `true` (the `keep` that's an
+    // argument to this function). We, however, don't want to keep either of
+    // those! Otherwise we'll get obscure resolve errors about locked versions.
+    //
+    // To solve this problem we apply similar logic to `keep(x) == false`
+    // dependencies. We look at all packages defined as path sources (aka on
+    // this filesystem). If any of them have dependency specifications which
+    // don't match our previous lockfile then we want to avoid locking them.
+    //
+    // For example let's say we had `serde = "0.1"` written in our lock file.
+    // When we later edit this to `serde = "0.1.3"` we don't want to lock serde
+    // at its old version, 0.1.1. Instead we want to allow it to update to
+    // `0.1.3` and update its own dependencies (like above).
+    //
+    // And finally, with all that in mind, hopefully this loop makes sense! If
+    // not, of course feel free to ask on IRC :). Like the function, this is
+    // directly targeted at solving #4127
+    let mut avoid_locking = HashSet::new();
+    for node in resolve.iter() {
+        if !keep(&node) {
+            add_deps(resolve, node, &mut avoid_locking);
+            continue
+        }
+        let path = match node.source_id().url().to_file_path() {
+            Ok(path) => path,
+            Err(_) => continue,
+        };
+        let pkg = match ws.load(&path.join("Cargo.toml")) {
+            Ok(p) => p,
+            Err(_) => continue,
+        };
+        for dep in resolve.deps_not_replaced(node) {
```

Comment:
So, here we are solving the case of "don't lock potentially changed dependencies". Another similar scenario might happen, I think, when the version of the pacakge itself changes (you literally change Cargo.toml from `[pacakge] name = foo version = 1` to `[package] name = foo version = 2`). So, perhaps we need something like 

```Rust
if pkg.package_id() != node {
  avoid_locking.insert(node);
}
```

is needed as well?